### PR TITLE
Dimensional rescaling of lengths and velocities

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
This PR includes widespread changes to the MOM6 code to allow for run-time
rescaling of horizontal lengths (including those in the MOM_grid_type),
horizontal velocities, transports and accelerations for dimensional consistency
testing.  All answers are bitwise identical, and there is only one minor
correction to the units of two diagnostics of accelerations that appear in the
available_diags files.  The list of commits with this PR includes:

- NOAA-GFDL/MOM6@170c7be Rescaled taux_bot and tauy_bot to [kg L Z T-2 m-3]
- NOAA-GFDL/MOM6@c0fddab Merge branch 'dev/gfdl' into transport_rescale
- NOAA-GFDL/MOM6@43636fa Merge branch 'dev/gfdl' into transport_rescale
- NOAA-GFDL/MOM6@4b62f77 Simplified a logical test in step_forward_MEKE
- NOAA-GFDL/MOM6@e3b8754 Rescaled MOM_open_boundary.F90 internal variables
- NOAA-GFDL/MOM6@d645d98 +Pass timestep to continuity in [T]
- NOAA-GFDL/MOM6@57eb24b +Pass velocities to write_energy in [L T-1]
- NOAA-GFDL/MOM6@29bb12b +initialize_vel routines pass velocities in [L T-1]
- NOAA-GFDL/MOM6@897fa67 +Change units of MOM_control_struct%u to [L T-1]
- NOAA-GFDL/MOM6@d0077ea +Pass velocities to calculate_diagnostic_fields in [L T-1]
- NOAA-GFDL/MOM6@80523af +Point MIS%u_prev to MOM_control_struct%u_prev
- NOAA-GFDL/MOM6@8d96f8d Introduced dt_in_T to step_MOM_dyn...
- NOAA-GFDL/MOM6@dd89a5d +Pass velocities to step_MOM_dyn_... in [L T-1]
- NOAA-GFDL/MOM6@3d594e0 MOM_dyn_split_RK2_CS%u_av in units of [L T-1]
- NOAA-GFDL/MOM6@7b0875f +Pass vels to radiation_open_bdry_conds in [L T-1]
- NOAA-GFDL/MOM6@670dcd9 Velocities in [L T-1] in step_MOM_dyn_split_RK2
- NOAA-GFDL/MOM6@887e36b +Pass velocities to vertvisc in [L T-1]
- NOAA-GFDL/MOM6@a0a46e8 +Rescaled ADp%du_dt_visc to units of [L T-2]
- NOAA-GFDL/MOM6@e90f16d +Recast vertvisc to work with velocities in [L T-1]
- NOAA-GFDL/MOM6@e15389d Code style modifications in wave_structure
- NOAA-GFDL/MOM6@b74e2d3 +Pass velocities to vertvisc_coef in [L T-1]
- NOAA-GFDL/MOM6@64f691c +Pass velocities to set_viscous_ML in [L T-1]
- NOAA-GFDL/MOM6@9d58553 +Pass velocities to btstep in [L T-1]
- NOAA-GFDL/MOM6@a7e5490 +Pass velocities to CorAdCalc in [L T-1]
- NOAA-GFDL/MOM6@ccdd508 +Pass velocities to continuity in [L T-1]
- NOAA-GFDL/MOM6@972f2cf +Pass velocities to step_MOM_thermo in [L T-1]
- NOAA-GFDL/MOM6@9fa6d39 +Pass velocities to set_viscous_BBL in [L T-1]
- NOAA-GFDL/MOM6@e35c52a +Added optional vel_scale argument to MOM_state_chksum_5arg
- NOAA-GFDL/MOM6@a2cbca6 +Pass velocities to diabatic in [L T-1]
- NOAA-GFDL/MOM6@e8c84a6 +Added thickness rescaling in MOM_state_stats
- NOAA-GFDL/MOM6@55aeaef +Pass velocities to bulkmixedlayer in [L T-1]
- NOAA-GFDL/MOM6@a8c41f5 +Pass velocities to set_diffusivity in [L T-1]
- NOAA-GFDL/MOM6@8cf1897 Rescaled velocities in MOM_energetic_PBL.F90
- NOAA-GFDL/MOM6@38ce6ac Rescaled variables in MOM_bulk_mixed_layer.F90
- NOAA-GFDL/MOM6@561cf95 +Pass internal wave speeds in [L T-1]
- NOAA-GFDL/MOM6@58cf1a9 +Rescaled CDp%uhGM diagnostics to [H L2 T-1]
- NOAA-GFDL/MOM6@54db6d3 Rescaled variables in MOM_internal_tides.F90
- NOAA-GFDL/MOM6@32aa6d5 Simplified scaling terms in find_uv_at_h
- NOAA-GFDL/MOM6@ce21fd0 Unit scaling term cleanup in set_viscous_BBL
- NOAA-GFDL/MOM6@c20db36 +Rescaled OBC%segment%grad_gradient to [m s-1 L-1]
- NOAA-GFDL/MOM6@b84e8ca +Rescaled OBC%segment%tangential_grad to [T-1]
- NOAA-GFDL/MOM6@6f5f24f Rescaled OBC%segment%normal_trans to units of [H L2 T-1]
- NOAA-GFDL/MOM6@d820888 +Rescaled OBC%segment%normal_vel to units of [L T-1]
- NOAA-GFDL/MOM6@d1267c9 Rescaled variables in MOM_CoriolisAdv.F90
- NOAA-GFDL/MOM6@cdbd580 +Return N2_u from calc_isoneutral_slopes in [T-2]
- NOAA-GFDL/MOM6@90bbde2 Correct rescaling diffu and diffv across restarts
- NOAA-GFDL/MOM6@a4ffe03 +Rescaled isoneutral & along layer tracer diffusion
- NOAA-GFDL/MOM6@e026e6e Rescaled variables in MOM_lateral_mixing_coeffs.F90
- NOAA-GFDL/MOM6@8067dfa +Rescaled VarMix%cg1 to units of [L T-1]
- NOAA-GFDL/MOM6@b4fa597 +Rescaled VarMix%SN_u and VarMix%SN_v to [T-1]
- NOAA-GFDL/MOM6@4031ab9 +Rescaled VarMix%L2u & VarMix%L2v to units of [L2]
- NOAA-GFDL/MOM6@c9f0b2a Rescaled variables in MOM_thickness_diffuse.F90
- NOAA-GFDL/MOM6@a7e832c Rescaled variables in MOM_mixedlayer_restrat.F90
- NOAA-GFDL/MOM6@133b03f +Rescaled G%dF_dx & G%dF_dy to units of [T-1 L-1]
- NOAA-GFDL/MOM6@f07e85f Rescaled internal variables in MOM_MEKE.F90
- NOAA-GFDL/MOM6@070c604 (*)Eliminated a bug in horizontal viscosity work
- NOAA-GFDL/MOM6@1ff9073 +Rescaled horizontal viscosity accelerations
- NOAA-GFDL/MOM6@b4df147 +Rescaled 9 elements of the MEKE_type
- NOAA-GFDL/MOM6@63ba91b Renamed u0 to Del2u in horizontal_viscosity
- NOAA-GFDL/MOM6@4a59fef +Rescaled div_xx_dx sent to calc_QG_Leith_viscosity
- NOAA-GFDL/MOM6@9438c73 +Rescaled ubtav returned by barotropic_get_tav
- NOAA-GFDL/MOM6@3689eb5 Rescaled internal variables in MOM_hor_visc.F90
- NOAA-GFDL/MOM6@6062801 Simplified rescaling in MOM_tracer_hor_diff.F90
- NOAA-GFDL/MOM6@b758eb4 Simplified rescaling in MOM_vert_friction.F90
- NOAA-GFDL/MOM6@57c2343 Simplified rescaling in MOM_set_diffusivity.F90
- NOAA-GFDL/MOM6@587e988 Simplified rescaling in MOM_thickness_diffuse.F90
- NOAA-GFDL/MOM6@77692a3 Simplified rescaling in MOM_hor_visc.F90
- NOAA-GFDL/MOM6@b72dd75 Simplified rescaling in MOM_MEKE.F90
- NOAA-GFDL/MOM6@7b11e00 Simplified rescaling in MOM_controlled_forcing.F90
- NOAA-GFDL/MOM6@37a570a Simplified rescaling in MOM_sum_output.F90
- NOAA-GFDL/MOM6@3130e36 +Rescaled the units of G%IdxCu and G%IdyCv
- NOAA-GFDL/MOM6@4bf27c6 +Rescaled the units of G%IdyCu and G%IdxCv
- NOAA-GFDL/MOM6@e1081a5 +Rescaled the units of G%IdxT and G%IdyT
- NOAA-GFDL/MOM6@dc6c85b +Rescaled the units of G%IdxBu and G%IdyBu
- NOAA-GFDL/MOM6@013b908 +Rescaled the units of G%dxBu and G%dyBu
- NOAA-GFDL/MOM6@9f3e8df +Rescaled the units of G%dyCu and G%dxCv
- NOAA-GFDL/MOM6@14ce1d4 +Rescaled the units of G%dxCu and G%dyCv
- NOAA-GFDL/MOM6@5ed5cdb +Rescaled the units of G%dyT and G%dxT
- NOAA-GFDL/MOM6@a673354 +Rescaled the units of G%dy_Cu and G%dx_Cv
- NOAA-GFDL/MOM6@9b25e9e +Add G%US and rescale the units of G%area arrays
- NOAA-GFDL/MOM6@0772696 Simplified some G%Iarea unit-scaling expressions
- NOAA-GFDL/MOM6@4cd8641 +Rescaled the units of G%Iarea arrays
- NOAA-GFDL/MOM6@0652259 Add parentheses around grid metric arrays
- NOAA-GFDL/MOM6@9d75ae9 +Rescaled continuity internal calculations
- NOAA-GFDL/MOM6@7e29087 Rescaled variables in advect_tracer
- NOAA-GFDL/MOM6@3db3085 +Relocated rescaling of pressure gradient terms
- NOAA-GFDL/MOM6@636a09d Rescaled pressure gradient accelerations
- NOAA-GFDL/MOM6@e49b1fd +Rescaled Coriolis accelerations
- NOAA-GFDL/MOM6@767c09d +Rescaled accelerations passed to btstep
- NOAA-GFDL/MOM6@20108b5 +Rescaled accelerations returned from btstep
- NOAA-GFDL/MOM6@89c9522 Dimensional rescaling inside continuity_PPM.F90
- NOAA-GFDL/MOM6@15438e8 +Eliminated uhbt_aux args from continuity_PPM
- NOAA-GFDL/MOM6@6c0f9f7 +Changed the units of uhbt and vhbt to [H L2 T-1]
- NOAA-GFDL/MOM6@0efc118 Corrected chksum scaling for vertvisc_CS%h_u
- NOAA-GFDL/MOM6@139c562 Cleanup of rescaling factors in energy diagnostics
- NOAA-GFDL/MOM6@40335f4 +Rescaled uh and vh to [H L2 T-1]
- NOAA-GFDL/MOM6@bc5d3b0 Merge branch 'dev/gfdl' into transport_rescale
- NOAA-GFDL/MOM6@eb9cbc2 +Rescaled uhtr and vhtr

### CODE REVIEW

**Comment by Hallberg-NOAA** [comment]
This PR is being tested with https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/8668.

File: src/diagnostics/MOM_wave_speed.F90 [hunk 1]
```diff
@@ -778,12 +779,12 @@ subroutine wave_speeds(h, tv, G, GV, US, nmodes, cn, CS, full_halos)
             do K=2,kc
               Igl(K) = 1.0/(gprime(K)*Hc(k)) ; Igu(K) = 1.0/(gprime(K)*Hc(k-1))
               z_int(K) = z_int(K-1) + Hc(k-1)
-              N2(K) = US%m_to_Z**2*gprime(K)/(0.5*(Hc(k)+Hc(k-1)))
+              ! N2(K) = US%m_to_Z**2*gprime(K)/(0.5*(Hc(k)+Hc(k-1)))
```
Comment (marshallward) [inline src/diagnostics/MOM_wave_speed.F90]: I think if we're going to keep these lines of commented code, then some explanation is needed, e.g. reproducibility, reliance on unimplemented features, etc.

**Comment by marshallward** [comment]
> @marshallward Should we hold of on this PR until after your testing is in place, or shall we put this through your tests?

I can test it offline, that's not a problem.

File: src/ice_shelf/MOM_ice_shelf_dynamics.F90 [hunk 2]
```diff
@@ -580,14 +580,16 @@ function ice_time_step_CFL(CS, ISS, G)
   real :: local_u_max, local_v_max
   integer :: i, j
 
-  min_ratio = 1.0e16 ! This is just an arbitrary large value.
+  min_ratio = 1.0e16 ! This is just an arbitrary large nondiensional value.
```
Comment (marshallward) [inline src/ice_shelf/MOM_ice_shelf_dynamics.F90]: A typo!!

**Comment by marshallward** [comment]
There are several changes in this PR to the OpenMP directives inside of `MOM_MEKE.F90`.  I am not sure if they are new or just regressions of previous changes by NCAR.  @nikizadehgfdl has previously noted other answer-changing issues with changes OpenMP, so my guess is that this should at least be addressed, either here or in another PR.

**Comment by marshallward** [comment]
BTW it seems Bob's PR has passed the new Travis test suite.

**Comment by marshallward** [comment]
@gustavo-marques Do these changes in the OpenMP directivse in `MOM_MEKE.F90` seem significant to you?   We are not currently testing any OpenMP loops in the CI.

**Comment by adcroft** [comment]
This requires a manual merge because it updates files in MOM6-examples. Handling now.

**Comment by gustavo-marques** [comment]
@marshallward We are not using OpenMP so these changes won't affect us. Thanks for checking.

**Pull request merged**